File: kakao/20년 인턴/동굴 탐험.py
def solution(n, path, order):
    b=[[] for i in range(n)]
    while path:s,e=path.pop();b[s].append(e);b[e].append(s)
    dest=set();s2e=[0]*n
    for s,e in order:dest.add(e);s2e[s]=e
    q=[0];v=[0]*n
    while q:
        node=q.pop();v[node]=1;e=s2e[node]
        if e:
            dest.remove(e)
            if v[e]:q.extend([i for i in b[e] if v[i]==0])
        elif node in dest:continue
        q.extend([i for i in b[node] if v[i]==0])
    return not dest

File: kakao/20년 인턴/동굴 탐험.cpp
#include <string>
#include <vector>
using namespace std;

int wait[200000],before[200000];
bool visit[200000];
vector<int> connect[200000];

void find(int num)
{
    if (visit[num])return;
    if(!visit[before[num]]){wait[before[num]]=num;return;}
    visit[num]=1;find(wait[num]);
    for(int i:connect[num])find(i);
}

bool solution(int n, vector<vector<int>> path, vector<vector<int>> order) {
    for(auto i:path)connect[i[0]].push_back(i[1]),connect[i[1]].push_back(i[0]);
    for(auto i:order)before[i[1]]=i[0];
    if(before[0])return false;
    visit[0]=1;
    for(int i:connect[0])find(i);
    for(int i=0;i<n;i++)if(!visit[i])return false;
    return true;
}